### PR TITLE
expose reparse functions, for 'fixing' an image proto to recognize all fields

### PR DESCRIPTION
**Pull request by jhump**

### CODE REVIEW

**Review by bufdev — APPROVED** [review]
Talked offline, Josh and I decided we're going to make reparse the default for `NewImageForProto`, with a `type NewImageForProtoOption func(*newImageForProtoOptions)` with `WithNoImageReparse` option that turns off reparsing for those who call `NewImageForProto` who really want to save that slight amount of time. This does not affect compiler speed, only image reads.

Approving assuming this is the change.

File: private/bufpkg/bufimage/bufimage.go [hunk 1]
```diff
@@ -167,13 +168,55 @@ func MergeImages(images ...Image) (Image, error) {
 	}
 }
 
+type NewImageForProtoOption func(*newImageForProtoOptions)
+
+type newImageForProtoOptions struct {
```
Comment (bufdev) [inline private/bufpkg/bufimage/bufimage.go:173]: private types always go below all public types, see style guide

File: private/bufpkg/bufimage/bufimage.go [hunk 2]
```diff
@@ -167,13 +168,55 @@ func MergeImages(images ...Image) (Image, error) {
...
+// WithNoReparse instructs NewImageForProto to skip the reparse step. The reparse
+// step is usually needed when unmarshalling the image from bytes. It reconstitutes
+// custom options, from unrecognized bytes to known extension fields.
+func WithNoReparse() NewImageForProtoOption {
+	return func(options *newImageForProtoOptions) {
+		options.noReparse = true
+	}
+}
+
+func reparseImageProto(protoImage *imagev1.Image) error {
```
Comment (bufdev) [inline private/bufpkg/bufimage/bufimage.go:186]: same comment

File: private/bufpkg/bufimage/bufimagetesting/bufimagetesting_test.go [hunk 3]
```diff
@@ -120,7 +122,7 @@ func TestBasic(t *testing.T) {
 		"b/b.proto",
 		"a/a.proto",
 		"a/b.proto",
-		"b/b.proto",
+		"b/a.proto",
```
Comment (jhump) [inline private/bufpkg/bufimage/bufimagetesting/bufimagetesting_test.go:125]: This was an import cycle -- b/b.proto imported b/b.proto.

Now that we try to build a resolver from the image, to resolve any unrecognized fields, this gets noticed in the 
NewImageForProto call.

File: private/bufpkg/bufimage/bufimagetesting/bufimagetesting_test.go [hunk 4]
```diff
@@ -451,11 +453,9 @@ func TestBasic(t *testing.T) {
 		},
 		newImage.Files(),
 	)
-	require.Equal(
-		t,
-		protoImage,
-		bufimage.ImageToProtoImage(newImage),
-	)
```
Comment (jhump) [inline private/bufpkg/bufimage/bufimagetesting/bufimagetesting_test.go:458]: This assertion was failing after the change. But the only thing I spotted were differences in internal unexported bits of the proto struct. So changing to do proper proto comparison fixed it.

File: private/bufpkg/bufimage/bufimage.go (hunk 1 above)
Comment (jhump) [reply private/bufpkg/bufimage/bufimage.go]: done

File: private/buf/bufwire/image_reader.go [hunk 5]
```diff
@@ -67,31 +67,14 @@ func (i *imageReader) GetImage(
...
 	}
 	protoImage := &imagev1.Image{}
+	var imageFromProtoOptions []bufimage.NewImageForProtoOption
 	switch imageEncoding := imageRef.ImageEncoding(); imageEncoding {
 	// we have to double parse due to custom options
 	// See https://github.com/golang/protobuf/issues/1123
 	// TODO: revisit
 	case buffetch.ImageEncodingBin:
-		firstProtoImage := &imagev1.Image{}
 		_, span := trace.StartSpan(ctx, "first_wire_unmarshal")
```
Comment (pkwarren) [inline private/buf/bufwire/image_reader.go:76]: nit: Should this span just be named `wire_unmarshal` now?

File: private/bufpkg/bufimage/bufimage.go [hunk 6]
```diff
@@ -250,6 +261,17 @@ func NewImageForCodeGeneratorRequest(request *pluginpb.CodeGeneratorRequest) (Im
 	)
 }
 
+type NewImageForProtoOption func(*newImageForProtoOptions)
```
Comment (pkwarren) [inline private/bufpkg/bufimage/bufimage.go:264]: Should add godoc strings for all exported types/methods.

File: private/pkg/protoencoding/protoencoding.go [hunk 7]
```diff
@@ -37,6 +37,17 @@ func NewResolver(fileDescriptors ...protodescriptor.FileDescriptor) (Resolver, e
 	return newResolver(fileDescriptors...)
 }
 
+// NewLazyResolver creates a new Resolver that is constructed from the given
+// descriptors only as needed, if invoked.
+//
+// If there is an error when constructing the resolver, it will be returned by all
+// method calls of the returned resolver.
+func NewLazyResolver(fileDescriptors ...protodescriptor.FileDescriptor) Resolver {
```
Comment (pkwarren) [inline private/pkg/protoencoding/protoencoding.go:45]: Is this an optimization to how things worked before? I noticed GetImage changed from `NewResolver` to `NewLazyResolver`.

**Review by pkwarren — APPROVED** [review]

File: private/bufpkg/bufimage/bufimage.go (hunk 6 above)
Comment (jhump) [inline private/bufpkg/bufimage/bufimage.go:264]:
```suggestion
// NewImageForProtoOption is an option for use with NewImageForProto.
type NewImageForProtoOption func(*newImageForProtoOptions)
```

**Pull request merged**